File: app/src/main/java/it/amonshore/comikkua/data/comics/ComicsDao.java
package it.amonshore.comikkua.data.comics;

import java.util.List;

import androidx.lifecycle.LiveData;
import androidx.paging.DataSource;
import androidx.paging.PagingSource;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

@Dao
public interface ComicsDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long insert(Comics comics);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    Long[] insert(Comics... comics);

    @Update
    int update(Comics comics);

    @Update
    int update(Comics... comics);

    @Query("UPDATE tComics SET removed = :removed WHERE id IN (:id)")
    @Transaction
    int updateRemoved(boolean removed, Long... id);

    @Query("UPDATE tComics SET removed = 0")
    @Transaction
    int undoRemoved();

    @Query("DELETE FROM tComics")
    int deleteAll();

    @Query("DELETE FROM tComics WHERE id = :id")
    int delete(long id);

    @Query("DELETE FROM tComics WHERE id in (:id)")
    @Transaction
    int delete(Long... id);

    @Query("DELETE FROM tComics WHERE removed = 1")
    @Transaction
    int deleteRemoved();

    @Query("SELECT * FROM tComics WHERE removed = 0 ORDER BY name COLLATE NOCASE ASC")
    LiveData<List<Comics>> getComics();

    @Query("SELECT * FROM tComics WHERE id = :id AND removed = 0")
    LiveData<Comics> getComics(long id);

    @Query("SELECT * FROM tComics WHERE name = :name AND removed = 0")
    LiveData<Comics> getComics(String name);

    @Query("SELECT * FROM tComics WHERE removed = 0 ORDER BY name COLLATE NOCASE ASC")
    List<Comics> getRawComics();

    @Query("SELECT * FROM tComics WHERE removed = 0 ORDER BY name COLLATE NOCASE ASC")
    @Transaction
    List<ComicsWithReleases> getRawComicsWithReleases();

    @Query("SELECT * FROM tComics WHERE removed = 0 ORDER BY name COLLATE NOCASE ASC")
    @Transaction
    LiveData<List<ComicsWithReleases>> getComicsWithReleases();

    @Query("SELECT * FROM tComics WHERE removed = 0 ORDER BY name COLLATE NOCASE ASC")
    @Transaction
    PagingSource<Integer, ComicsWithReleases> getComicsWithReleasesPagingSource();

    @Query("SELECT * FROM tComics WHERE removed = 0 AND " +
            "(name LIKE :likeName OR publisher LIKE :likeName OR authors LIKE :likeName OR notes LIKE :likeName) " +
            "ORDER BY name COLLATE NOCASE ASC")
    @Transaction
    PagingSource<Integer, ComicsWithReleases> getComicsWithReleasesPagingSource(String likeName);

    @Query("SELECT * FROM tComics WHERE name = :name AND removed = 0")
    @Transaction
    LiveData<List<ComicsWithReleases>> getComicsWithReleasesByName(String name);

    @Query("SELECT * FROM tComics WHERE id = :id AND removed = 0")
    @Transaction
    LiveData<ComicsWithReleases> getComicsWithReleases(long id);

    @Query("SELECT distinct(publisher) FROM tComics WHERE publisher IS NOT NULL AND publisher <> '' ORDER BY publisher")
    LiveData<List<String>> getPublishers();

    @Query("SELECT distinct(authors) FROM tComics WHERE authors IS NOT NULL AND authors <> '' ORDER BY authors")
    LiveData<List<String>> getAuthors();

    @Query("SELECT distinct(name) FROM tComics ORDER BY name COLLATE NOCASE ASC")
    LiveData<List<String>> getComicsName();
}

File: app/src/main/java/it/amonshore/comikkua/data/comics/ComicsDao.kt
package it.amonshore.comikkua.data.comics

import androidx.paging.PagingSource
import androidx.room.*
import kotlinx.coroutines.flow.Flow

@Dao
interface ComicsDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(comics: Comics)

    @Upsert
    suspend fun upsert(comics: Comics)

    @Query("DELETE FROM tComics WHERE removed = 1 AND tag = :tag")
    @Transaction
    suspend fun deleteRemoved(tag: String)

    @Query("UPDATE tComics SET removed = 0 WHERE tag = :tag")
    @Transaction
    suspend fun undoRemoved(tag: String)

    @Query("UPDATE tComics SET removed = 1, tag = :tag WHERE id IN (:ids)")
    @Transaction
    suspend fun markedAsRemoved(ids: List<Long>, tag: String): Int

    @Query("SELECT * FROM tComics WHERE name = :name COLLATE NOCASE AND removed = 0 AND selected = 1")
    suspend fun getComicsByName(name: String): Comics?

    @Query("SELECT * FROM tComics WHERE removed = 0 AND selected = 1 ORDER BY name COLLATE NOCASE ASC")
    @Transaction
    suspend fun getAllComicsWithReleases(): List<ComicsWithReleases>

    @Query("SELECT * FROM tComics WHERE id = :id AND removed = 0 AND selected = 1")
    @Transaction
    suspend fun getComicsWithReleases(id: Long): ComicsWithReleases

    @Query("SELECT id FROM tComics WHERE removed = 1 AND tag = :tag")
    suspend fun getRemovedComicsIds(tag: String): List<Long>

    @Query("SELECT * FROM tComics WHERE id = :id AND removed = 0 AND selected = 1")
    @Transaction
    fun getComicsWithReleasesFlow(id: Long): Flow<ComicsWithReleases>

    @Query("SELECT * FROM tComics WHERE removed = 0 AND selected = 1 ORDER BY name COLLATE NOCASE ASC")
    @Transaction
    fun getComicsWithReleasesPagingSource(): PagingSource<Int, ComicsWithReleases>

    @Query(
        """
            SELECT * 
            FROM tComics 
            WHERE 
            removed = 0 AND selected = 1 
            AND (name LIKE :like 
                OR publisher LIKE :like 
                OR authors LIKE :like 
                OR notes LIKE :like)
            ORDER BY name COLLATE NOCASE ASC
            """
    )
    @Transaction
    fun getComicsWithReleasesPagingSource(like: String): PagingSource<Int, ComicsWithReleases>

    @Query(
        """
        SELECT distinct(publisher) 
        FROM tComics 
        WHERE publisher IS NOT NULL AND publisher <> '' 
        ORDER BY publisher
        """
    )
    suspend fun getPublishers(): List<String>

    @Query(
        """
        SELECT distinct(authors) 
        FROM tComics 
        WHERE authors IS NOT NULL AND authors <> '' 
        ORDER BY authors
        """
    )
    suspend fun getAuthors(): List<String>

}